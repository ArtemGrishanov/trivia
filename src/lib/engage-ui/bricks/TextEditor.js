import React from 'react'
import { setComponentProps } from '../../remix'

// https://github.com/zenoamaro/react-quill
import ReactQuill from 'react-quill'
import 'react-quill/dist/quill.snow.css'
import '../style/rmx-text_editor.css'

const Parchment = ReactQuill.Quill.import('parchment')
// line-height
const lineHeightConfig = {
    scope: Parchment.Scope.INLINE,
    whitelist: ['1', '1.2', '1.5', '1.6', '1.8', '2', '2.4', '2.8'],
}
const lineHeightClass = new Parchment.Attributor.Class('lineheight', 'ql-line-height', lineHeightConfig)
const lineHeightStyle = new Parchment.Attributor.Style('lineheight', 'line-height', lineHeightConfig)
Parchment.register(lineHeightClass)
Parchment.register(lineHeightStyle)
// font-size
const fontSizeConfig = {
    scope: Parchment.Scope.INLINE,
    whitelist: [
        '8px',
        '9px',
        '10px',
        '11px',
        '12px',
        '14px',
        '16px',
        '18px',
        '20px',
        '24px',
        '30px',
        '36px',
        '48px',
        '60px',
        '72px',
        '96px',
    ],
}
const fontSizeClass = new Parchment.Attributor.Class('fontsize', 'ql-font-size', fontSizeConfig)
const fontSizeStyle = new Parchment.Attributor.Style('fontsize', 'font-size', fontSizeConfig)
Parchment.register(fontSizeClass)
Parchment.register(fontSizeStyle)

const // To add a new font:
    // 1. Add font into this array
    // 2. Add classes as UpperCamelCase in rmx-text_editor.css
    fonts = [
        'Roboto',
        'Roboto Condensed',
        'Open Sans',
        'Open Sans Condensed',
        'Lato',
        'Montserrat',
        'Oswald',
        'Merriweather',
        'Ubuntu',
        'Lobster',
        'Pacifico',
        'Vollkorn',
        'Cuprum',
        'Alegreya Sans',
        'Russo One',
        'Playfair Display SC',
        'Alice',
        'Press Start 2P',
        'Bad Script',
        'Yeseva One',
        'Marmelad',
        'Rubik Mono One',
        'Raleway',
        'Roboto Slab',
        'Lora',
        'Seymour One',
        'Cormorant SC',
        'Literata',
        'Spectral',
        'Alegreya',
        'EB Garamond',
        'Bitter',
        'PT Serif',
        'Noto Sans',
    ],
    importedFonts = {
        // 'Ubuntu': true | false
        // ...
    }

// https://github.com/zenoamaro/react-quill/issues/273
const Font = ReactQuill.Quill.import('formats/font')
Font.whitelist = fonts.map(f => f.replace(/ /g, ''))
ReactQuill.Quill.register(Font, true)

export function getStylesheetLink(family) {
    return `https://fonts.googleapis.com/css?family=${family}`
}

export function addFont(family) {
    if (!importedFonts[family]) {
        const link = document.createElement('link')
        link.href = getStylesheetLink(family)
        link.rel = 'stylesheet'
        document.getElementsByTagName('head')[0].append(link)
        importedFonts[family] = true
    }
}

export class TextEditor extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            html: props.text,
        }
        this.handleChange = this.handleChange.bind(this)
        this.contentRef = React.createRef()
        this.measureInterval = null
        this.measuredRect = {}
    }

    handleChange(html) {
        this.setState({ html })
        // import font which is used in app
        fonts.forEach((f, i) => {
            if (html.includes(`ql-font-${Font.whitelist[i]}`)) {
                addFont(f)
            }
        })
    }

    modules = {
        toolbar: [
            [{ font: Font.whitelist }, { fontsize: fontSizeConfig.whitelist }],
            ['bold', 'italic', 'underline', { color: [] }, { background: [] }],
            [{ align: [] }, { lineheight: lineHeightConfig.whitelist }],
            ['link', 'clean'],
        ],
    }

    // 1. when you open editor first time in edit mode import all fonts
    // 2. always - check which fonts is used and add styles dynamically
    render() {
        const { readOnly, className = '' } = this.props
        return (
            <div
                className={`rmx-text-editor ${readOnly ? '__readonly' : ''}${className.length ? ` ${className}` : ''}`}
                ref={this.contentRef}
            >
                <ReactQuill
                    readOnly={readOnly}
                    modules={this.modules}
                    value={this.state.html}
                    onChange={this.handleChange}
                />
            </div>
        )
    }

    componentDidMount() {
        if (this.props.getContentRect) {
            if (this.contentRef && this.contentRef.current) {
                //TODO попробовать react-size-me в будущем как, возможно, более производительное решение. Хотя этот код измерения текста запускается очень редко и не стоит добавлять лишние библиотеки
                this.measureInterval = setInterval(() => {
                    const { width, height } = this.contentRef.current.getBoundingClientRect()
                    //TODO эксперимент - нужна только высота. Пробую убрать ширину.
                    if (/*this.measuredRect.width !== width ||*/ this.measuredRect.height !== height) {
                        this.measuredRect = { /*width,*/ height }
                        this.props.getContentRect({ /*width,*/ height })
                    }
                }, 400)
            }
        }
    }

    componentWillUnmount() {
        if (this.measureInterval) {
            clearInterval(this.measureInterval)
        }
    }

    componentDidUpdate({ readOnly: prevReadOnly }) {
        const { readOnly, parentId } = this.props
        if (!readOnly) {
            // import all fonts in edit mode
            fonts.forEach(f => addFont(f))
            const { getContentRect } = this.props
            if (getContentRect && this.contentRef && this.contentRef?.current) {
                getContentRect(this.contentRef.current.getBoundingClientRect())
            }
        }
        if (!prevReadOnly && readOnly) {
            // выход из режима ввода текста - делаем сохранение в remix
            setComponentProps(
                {
                    id: this.props.parentId,
                    text: this.state.html,
                },
                { putStateHistory: true },
            )
        }
    }
}

export default TextEditor
