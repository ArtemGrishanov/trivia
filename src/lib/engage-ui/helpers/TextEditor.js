import React from 'react'

// https://github.com/zenoamaro/react-quill
import ReactQuill from 'react-quill'
import 'react-quill/dist/quill.snow.css'

class TextEditor extends React.Component {

    static getDerivedStateFromProps(props, state) {
        return {
            ...state,
            // props.onChange - значит родитель извне управляет текстом
            stateText: props.onChange ? props.text: state.stateText
        }
    }

    constructor(props) {
        super(props);
        this.state = {
            stateText: props.text
        }
        this.handleChange = this.handleChange.bind(this)
    }

    handleChange(value) {
        if (this.props.onChange) {
            // отправить изменение вверх родительскому компоненту
            // он позаботится о сохранении значения и вернет его через props.text
            this.props.onChange(value);
        }
        else {
            this.setState({ stateText: value })
        }
    }

    modules = {
        toolbar: [
            ['bold', 'italic', 'underline'],                  // toggled buttons
            [{ 'size': ['small', false, 'large', 'huge'] }],  // custom dropdown
            [{ 'color': [] }, { 'background': [] }],          // dropdown with defaults from theme
            [{ 'font': [] }],
            [{ 'align': [] }],
            ['link', 'clean']
        ]
    }

    render() {
        return (
            <div className={`rmx-text-editor ${this.props.readOnly ? '__readonly' : ''}`}>
                <ReactQuill readOnly={this.props.readOnly}
                            modules={this.modules}
                            value={this.state.stateText}
                            onChange={this.handleChange} />
            </div>
        )
    }

    componentDidMount() {
    }

    componentDidUpdate() {
    }
}

export default TextEditor