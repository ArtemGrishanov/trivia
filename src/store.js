import { createStore, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'

import reducer from './reducer'
import eventSaga from './lib/remix/sagas/triggerExecutor'
import diffMiddleware from './lib/remix/middleware/diff.js'

const sagaMiddleware = createSagaMiddleware()
const store = createStore(reducer, applyMiddleware(sagaMiddleware, diffMiddleware))

sagaMiddleware.run(eventSaga)

window.store = store // for debug: inspect storage state in browser console

export default store

/**
 * Simple middleware example
 *
 * @param {*} store
 */
function logger(store) {
  return function (next) {
    return function (action) {
      console.log('dispatching', action)
      let result = next(action)
      console.log('next state', store.getState())
      return result
    }
  }
}
