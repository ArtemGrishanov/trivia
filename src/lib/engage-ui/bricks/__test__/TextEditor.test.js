import React from 'react'
import { shallow } from 'enzyme'

import { TextEditor, getStylesheetLink, addFont } from './../TextEditor'

import store from './../../../../store'
import Remix from './../../../../lib/remix'

describe('<TextEditor />', () => {
    let wrapper
    let props

    beforeAll(() => {
        props = {
            readOnly: true,
            text: `
                <p class="ql-align-center">
                    <strong class="ql-size-huge" style="color: rgb(255, 255, 255);">ARE YOU GOOD AT GEOGRAPHY?</strong>
                </p>`,
        }
        Remix.setStore(store)
    })

    it('should match <TextEditor /> snapshot', () => {
        wrapper = shallow(<TextEditor {...props} />)
        expect(wrapper).toMatchSnapshot()
    })

    it('should match <TextEditor /> snapshot in readOnly mode', () => {
        wrapper = shallow(<TextEditor {...props} readOnly={false} />)
        expect(wrapper).toMatchSnapshot()
    })

    it('should change font familly', () => {
        const newText = `
            <p class="ql-align-center">
                <strong class="ql-size-huge ql-font-Yeseva-One" style="color: rgb(255, 255, 255);">ARE YOU GOOD AT GEOGRAPHY?</strong>
            </p>`
        wrapper = shallow(<TextEditor {...props} />)
        const instance = wrapper.instance()
        instance.handleChange(newText)
        expect(wrapper.state().html).toEqual(newText)
        expect(wrapper.find('Quill').props().value).toEqual(newText)
    })
})

describe('getStylesheetLink', () => {
    it('should be equal fontFamily', () => {
        const fontFamily = 'SupaFont'
        const link = `https://fonts.googleapis.com/css?family=${fontFamily}`
        expect(getStylesheetLink(fontFamily)).toEqual(link)
    })
})

describe('addFont', () => {
    it('should add font', () => {
        const fontFamily = 'SupaFont'
        const link = `<link href="https://fonts.googleapis.com/css?family=${fontFamily}" rel="stylesheet">`
        addFont(fontFamily)
        expect(document.getElementsByTagName('head')[0].innerHTML).toEqual(link)
    })
})
