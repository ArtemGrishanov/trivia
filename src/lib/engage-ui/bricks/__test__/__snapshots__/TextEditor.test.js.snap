// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextEditor /> should match <TextEditor /> snapshot 1`] = `
<div
  className="rmx-text-editor __readonly"
>
  <Quill
    modules={
      Object {
        "toolbar": Array [
          Array [
            "bold",
            "italic",
            "underline",
          ],
          Array [
            Object {
              "fontsize": Array [
                "10px",
                "12px",
                "14px",
                "16px",
                "18px",
                "20px",
                "22px",
                "24px",
              ],
            },
          ],
          Array [
            Object {
              "lineheight": Array [
                "1.0",
                "1.2",
                "1.5",
                "1.6",
                "1.8",
                "2.0",
                "2.4",
                "2.8",
                "3.0",
                "4.0",
                "5.0",
              ],
            },
          ],
          Array [
            Object {
              "color": Array [],
            },
            Object {
              "background": Array [],
            },
          ],
          Array [
            Object {
              "font": Array [
                "Roboto",
                "RobotoCondensed",
                "OpenSans",
                "Lato",
                "Montserrat",
                "Oswald",
                "Merriweather",
                "Ubuntu",
                "Lobster",
                "Pacifico",
                "Vollkorn",
                "Cuprum",
                "AlegreyaSans",
                "RussoOne",
                "PlayfairDisplaySC",
                "Alice",
                "PressStart2P",
                "BadScript",
                "YesevaOne",
                "Marmelad",
                "RubikMonoOne",
              ],
            },
          ],
          Array [
            Object {
              "align": Array [],
            },
          ],
          Array [
            "link",
            "clean",
          ],
        ],
      }
    }
    onChange={[Function]}
    readOnly={true}
    theme="snow"
    value="
                <p class=\\"ql-align-center\\">
                    <strong class=\\"ql-size-huge\\" style=\\"color: rgb(255, 255, 255);\\">ARE YOU GOOD AT GEOGRAPHY?</strong>
                </p>"
  />
</div>
`;

exports[`<TextEditor /> should match <TextEditor /> snapshot in readOnly mode 1`] = `
<div
  className="rmx-text-editor "
>
  <Quill
    modules={
      Object {
        "toolbar": Array [
          Array [
            "bold",
            "italic",
            "underline",
          ],
          Array [
            Object {
              "fontsize": Array [
                "10px",
                "12px",
                "14px",
                "16px",
                "18px",
                "20px",
                "22px",
                "24px",
              ],
            },
          ],
          Array [
            Object {
              "lineheight": Array [
                "1.0",
                "1.2",
                "1.5",
                "1.6",
                "1.8",
                "2.0",
                "2.4",
                "2.8",
                "3.0",
                "4.0",
                "5.0",
              ],
            },
          ],
          Array [
            Object {
              "color": Array [],
            },
            Object {
              "background": Array [],
            },
          ],
          Array [
            Object {
              "font": Array [
                "Roboto",
                "RobotoCondensed",
                "OpenSans",
                "Lato",
                "Montserrat",
                "Oswald",
                "Merriweather",
                "Ubuntu",
                "Lobster",
                "Pacifico",
                "Vollkorn",
                "Cuprum",
                "AlegreyaSans",
                "RussoOne",
                "PlayfairDisplaySC",
                "Alice",
                "PressStart2P",
                "BadScript",
                "YesevaOne",
                "Marmelad",
                "RubikMonoOne",
              ],
            },
          ],
          Array [
            Object {
              "align": Array [],
            },
          ],
          Array [
            "link",
            "clean",
          ],
        ],
      }
    }
    onChange={[Function]}
    readOnly={false}
    theme="snow"
    value="
                <p class=\\"ql-align-center\\">
                    <strong class=\\"ql-size-huge\\" style=\\"color: rgb(255, 255, 255);\\">ARE YOU GOOD AT GEOGRAPHY?</strong>
                </p>"
  />
</div>
`;
